qutip/core/metrics.py

```diff
@@ -105,12 +108,104 @@ def fidelity(A, B):
 
 
-def process_fidelity(U1, U2, normalize=True):
+def _process_fidelity_to_id(oper):
+    """
+    Internal function returning the process fidelity of a quantum channel
+    to the identity quantum channel.
+    Parameters
+    ----------
+    oper : :class:`qutip.Qobj`/list
+        A unitary operator, or a superoperator in supermatrix, Choi or
+        chi-matrix form, or a list of Kraus operators
+    Returns
+    -------
+    fid : float
     """
-    Calculate the process fidelity given two process operators.
+    msg = 'The process fidelity to identity is only defined for ' \
+          'dimension preserving channels.'
+    if isinstance(oper, list):  # oper is a list of Kraus operators
+        d = oper[0].shape[0]
+        if oper[0].shape[1] != d:
+            raise TypeError(msg)
+        return np.sum([np.abs(k.tr()) ** 2 for k in oper]) / d ** 2
+    elif oper.type == 'oper':  # interpret as unitary
+        d = oper.shape[0]
+        if oper.shape[1] != d:
+            raise TypeError(msg)
+        return np.abs(oper.tr()) ** 2 / d ** 2
+    elif oper.type == 'super':
+        d = np.prod(oper.dims[0][0])
+        if np.prod(oper.dims[1][0]) != d:
+            raise TypeError(msg)
+        if oper.superrep == 'chi':
+            return oper[0, 0].real / d ** 2
+        else:  # oper.superrep is either 'super' or 'choi':
+            return to_super(oper).tr().real / d ** 2
+
+
+def _kraus_or_qobj_to_choi(oper):
+    if isinstance(oper, list):
+        return kraus_to_choi(oper)
+    else:
+        return to_choi(oper)
+
+
+def process_fidelity(oper, target=None):
     """
-    out = (U1 * U2).tr()
-    return out / (U1.tr() * U2.tr()) if normalize else out
+    Returns the process fidelity of a quantum channel to the target
+    channel, or to the identity channel if no target is given.
+    The process fidelity between two channels is defined as the state
+    fidelity between their normalized Choi matrices.
+    Parameters
+    ----------
+    oper : :class:`qutip.Qobj`/list
+        A unitary operator, or a superoperator in supermatrix, Choi or
+        chi-matrix form, or a list of Kraus operators
+    target : :class:`qutip.Qobj`/list
+        A unitary operator, or a superoperator in supermatrix, Choi or
+        chi-matrix form, or a list of Kraus operators
+    Returns
+    -------
+    fid : float
+        Process fidelity between oper and target, or between oper and identity.
+    Notes
+    -----
+    See, for example: A. Gilchrist, N.K. Langford, M.A. Nielsen,
+    Phys. Rev. A 71, 062310 (2005).
+    The definition of state fidelity that the process fidelity is based on
+    is the one from R. Jozsa, Journal of Modern Optics, 41:12, 2315 (1994).
+    It is the square of the one implemented in
+    :func:`qutip.metrics.fidelity` which follows Nielsen & Chuang,
+    "Quantum Computation and Quantum Information"
+    """
+    warnings.warn("Caution: Behavior of process_fidelity changed in Qutip 5.0")
```

Comment:
My vote would be to remove this warning and instead write a small paragraph in the docstring describing what changed between 4.7 and 5.0.

Reply:
This is a new function that replace an old one with only the name and signature in common. 

I don't expect any users to read the documentation of every function they use before trying their code in v5. Most will just try their old code, get different results and have hard time finding where it come from. Or worst, don't realize that they got a different result until much later...
So we should have a loud message that this changed, something that will make sure that any users using it will know to expect a new, unrelated result for that function call from now on.

Another option would be to make this a new function and deprecate the other one. This will remove any risk of confusion.

Reply:
My guess would be that not many people used the old `process_fidelity`, I don't think it was correct. 
`average_gate_fidelity` is closely related and probably more common.

As a next step, I'm planning to send a PR for `average_gate_fidelity`, reducing it to `process_fidelity`. (The current implementation of `average_gate_fidelity` is correct, but not efficient for some `oper.superrep`s.) That change wouldn't be user facing - so if we leave the warning, users of `average_gate_fidelity` would get a warning from `process_fidelity` even though nothing has changed (unless we introduce another private "warning-free" version of process fidelity but that seems complicated).

For those two reasons, I'd also suggest to remove the warning, but it's up to you.

Reply:
If it was completely wrong before, then we definitely don't need the warning -- and we've fixed such "this function does completely the wrong thing" in bugfix releases before without a warning.

If we do keep the warning, it should definitely say something more helpful than "something changed" which immediately causes the reader to wonder "what changed?".

@Ericgig Thoughts?

Reply:
I looked at when it was created , #130, the commit messages in the same PR mention GRAPE, so it seems to be a definition used in quantum control. `fidcomp.py` seems to have a similar function with `U1` being `_onto_evo` and `U2` being `_fwd_evo`. So it never mean to a fidelity between two operators but a fidelity contribution of one time slice of an iteration of GRAPE...

Since it was clearly not a mistake, it's not a sign error but a totally different formulae, I expected this was the right behaviour for someone's research and wanted to be safe.  But now that it seems to be part of an old implementation of quantum control that as been removed since. I don't see any issue with replacing it with only a note in the changelogs.

Reply:
@Ericgig Thank you for digging into the history!

@fhopfmueller If you will make the last change to remove the warning and add a note we can merge this, I think.

Reply:
Thanks a lot @Ericgig , that makes a lot of sense now!